### ISSUE TITLE
Teensy 4.1 is coming.

### ISSUE THREAD

**Issue by phil-barrett** [terjeio/grblHAL#22]
They have locked down the design and are in the process of qualifying suppliers so it's probably 2-3 months away. Of course the Corona virus pandemic may cause disruptions. It will use the IMXRT1062 and be in the same form factor as the teensy 3.6. All the T4.0 back side pins will be moved to TH pins so it's very likely that the Teensy4 GRBLHal will run without changes. The interesting thing is it appears it will support ethernet though it's a little unclear exactly what that means. Hopefully not via backside pins.

I'm hoping to get on their beta list though probably will not happen because you have to be a fairly active participant on the their forum.  Hoping my breakout board for T4.0 will change their minds.  By the way, my BoB seems to be working fine. Still a few things to test  (0-10V generation) but very optimistic.

**Comment by terjeio** [comment]
Interesting. Running FreeRTOS and LwIP for Ethernet...?

Great that your BoB seem to be fine. Have you tested the EEPROM yet?

I am still struggling with the USB stuff, by now I suspect my USB hub is also part of the problems I have. No way I could get MKRZERO to behave with my development machine, but it seems to run ok with the laptop. More testing tomorrow...

**Comment by phil-barrett** [comment]
Have not tested eeprom yet. Are there any issues using other that the 24LC16B? I have a 24FC256 which has the same pinout and, I believe, same command set.  

Paul didn't post a lot of detail and his ethernet support comment was in response to a question. 

Yeah, I've never met a USB hub that didn't cause some sort of problems. By the way, after looking at the Due and doing a preliminary layout, I probably won't do a BoB for it and instead considering one for the ZERO.  I had forgotten how much I hate the big Arduino form factors, especially the Mega but even the Uno is like an ugly, bitchy sister. What a mess.

**Comment by terjeio** [comment]
24LC16B is **not** command compatible with larger EEPROMS, it picks the page to be accessed from the I2C address and uses a single byte address within the page. Larger EEPROMS responds only to a single I2C address and uses two address bytes to access data.

It is on my todo list to add generic support for larger sizes, I have 256Kbit on hand for testing - but I need to make a breakout board first since it is a SMD device. First step is already done, as I've moved EEPROM code to the plugins folder.

FYI, I tested the MKRZERO with the 10hrs job overnight and it completed ok. However, I still have some issues remaining with the sender - maybe I need to rewrite the state handlers. A complicating factor is that many grbHAL drivers supports manual tool change. It is possible to jog when a program is halted for a tool change - this to allow for manual re-zeroing of Z. It should also be possible to execute a probe sequence in this state, all this without upsetting the senders internal state with regard to the running (but halted) program.

I am no fan of the Arduino stuff either, especially the lack of a debugger is a big pain. I do not use the IDE editor either...

I've attached a PCB drill file with tool changes if you want to check out how that works:

[Prototype-drill.nc.txt](https://github.com/terjeio/grblHAL/files/4261572/Prototype-drill.nc.txt)

**Comment by terjeio** [comment]
I've added support for larger EEPROMs, use #define EEPROM_ENABLE 2. EEPROM plugin code needs to be updated for that.

Also, I have added a second option for serial over native USB (PJRC driver) - but this still fails for me.

**Comment by einencool** [comment]
Today I ordered a Teensy 4.1
It should be delivered this month, when I can trust the webshop *g*
Is it also possible to use the "Ethernet" Shield or something like that? My laptop has only 2 USB Ports and one of them is having trouble with a stable connection...
@phil-barrett 
I hope I can test your new BoB in the next month :-)

**Comment by phil-barrett** [comment]
I ordered one too.  I have the layout done and have reworked the relay section to support both directly driving relays and TTL signals for things like SSRs and those Chinese relay boards.  I also cleaned up the stepper enable signals and support up to 5 independent enables. Added direct control of dust extraction.  It can be triggered off the spindle control or via a header with a switch attached.

My plan is to test the T4.1 via solderless breadboard and validate the BoB design, then manufacture the PCB and test. So, you will probably see it sometime in June.

**Comment by phil-barrett** [comment]
As to ethernet, I don't think it will ready to use right away so I haven't added it to the BoB but am planning a version that connects directly to the T4.1 header pins. with the magjack connector - essentially integrating the OSHPark adaptor mentioned on the PJRC page.

**Comment by terjeio** [comment]
If a [lwIP](http://savannah.nongnu.org/projects/lwip/) driver is available then it should be fairly easy to add ethernet as I have made a [plugin](https://github.com/terjeio/grblHAL/tree/master/plugins/networking) that is currently used for 3 different controllers.

**Comment by phil-barrett** [comment]
Well then, it sounds like a possibility.  https://forum.pjrc.com/threads/60532-Teensy-4-1-Beta-Test?p=237096&viewfull=1#post237096

**Comment by phil-barrett** [comment]
I think that I will just integrate the MagJack onto the first version of the T4.1 BoB.  It looks pretty easy.  Should have a design done later today. My T4.1 has shipped so I expect it on Thursday. I'm looking for a supplier of the MagJack as Mouser, my preferred disti, is out of stock and won't have any until June. I probably just send the board off to the fab without any ethernet testing.

I have chosen a naming convention that goes <processor><connection><num axis>BB so, the 4 axis T4.0 board is T4U4XBB and this new one will be T41E5XBB.  I am also considering getting a version with all the SMD components manufactured, sort of an "un-kit". Just add the TH components.  Thinking it would price in the $25 USD range. I'd still put the basic PCB design into open source, though.

**Comment by einencool** [comment]
Wow, you both are amazing 👍 
I am so thankful that you are able to set up such projects

**Comment by terjeio** [comment]
> to support both directly driving relays and TTL signals for things like SSRs and those Chinese relay boards.

Are there room/pins available for two inputs from an encoder? For lathe spindle sync (threading).

**Comment by phil-barrett** [comment]
Yes, there are pins available. I'll take a look.  Are there any specific needs or issues I should be aware of?   What commercial BoB/Controllers should I take a look at?

**Comment by terjeio** [comment]
> Are there any specific needs or issues I should be aware of? 

Preferably one pin should be routable to a counter clock input, I'll check the docs tomorrow.

> What commercial BoB/Controllers should I take a look at?

I do not know, and I do not know if it is relevant. If at some point tapping is implemented perhaps 3 pins are needed so that spindle direction can be determined too... Quadrature encoder input + index input.

> I assume Linux CNC is the model you are following.

Yes, that is the basic spec I try to follow. But not every detail is possible to implement in firmware.

If I can get spindle sync to work properly and port that to T4.1 then IMO this may become a good selling point for your board. For now I have one remaining issue I am struggling with which baffels me big time - with luck I just spotted where it originates.  More testing tomorrow. I am using a PID loop to control the position, and if I've coded this right then the accuracy in the control loop in the  ±20 µm range...

**Comment by phil-barrett** [comment]
Thanks.  It sounds like it's not really direction that is needed but an index pulse.  Anyway, good news, I still have 4 pins open and they all are interrupt capable. One issue - I am running out of connector space on the edge.  Would a pin header be acceptable instead of a screw terminal? I

**Comment by terjeio** [comment]
A pin header is acceptable, 5 pins? 3V3, PHASEB, PHASEA, INDEX and GND in that order?

It seems that you are free to use the pins of your choice, from chapter 60 in the reference manual:

> The intended application of this module is to provide a flexible crossbar switch function
that allows any input (typically from external GPIO or internal module outputs) to be
connected to any output (typically to external GPIO or internal module inputs) under user
control. This is used to allow user configuration of data paths between internal modules
and between internal modules and GPIO.

**Comment by phil-barrett** [comment]
No problem.  

Will do after I'm done fighting Eagle's differential pair routing. How well does KiCad do that? Maybe it is time to make the move.

**Comment by phil-barrett** [comment]
Questions as I look into it more.  These pins need to be isolated - via optoisolators. What level of RPM do you expect? The optos I'm using are good to about 25KHz. Can you give me a link to the data sheet for your encoder?

Also, you mention 3.3V - is this enough for most quad encoders?  I've seen some that say 5V.  Perhaps I should make the supply voltage configurable.

**Comment by terjeio** [comment]
> How well does KiCad do that?

I do not know since I have not done any differential routing. Google "kicad differential routing" for some leads like [this](https://forum.kicad.info/t/differential-pair-routing/12672). Forum members are friendly so they may help out if you encounter problems.

> Can you give me a link to the data sheet for your encoder?

It is [homemade](https://hackaday.io/project/165248-mini-lathe-emco-compact-5-cnc-conversion) - optical, 120 PPR, no quadrature as I do not need that.

My thinking is that the port should be at 3.3V and level shifters should be added by the end users if required. IMO bit dangerous to put 5V there (configurable is ok?) as it could be easy to fry the processor by the unaware. I have not checked the dynamic specs for the ENC peripheral, but I guess it can handle input pulses in the MHz range, the TM4C123 QEI peripheral surely can.

**Comment by phil-barrett** [comment]
I'm using optoisolators, there's no danger to the processor. Could be to the encoder if the voltage is misconfigured, though. Not sure how to prevent pilot error.  

If your motor is running your encoder at 3000 RPM, then that is 6KHz so my LiteOn optos (LTV 8x6) will do just fine. I've seen encoders that have upwards of 1000 PPR which would overwhelm them. Fortunately most applications don't have super high RPM.

Is your encoder able to drive 5 mA? My optos will work with less but I like a little extra margin.

**Comment by terjeio** [comment]
My encoder uses optocouplers already (slotted, there are holes in the disk for creating pulses). So I do not worry about that. BTW I have an open drain FET on the output capable of sinking 200mA. 

I've seen comments elsewhere from people CNCing their lathes and using encoders supplying up to 4096 pulses/rev. AFAIK some (most?) are optical, at least quadrature encoders - and some uses Hall elements. I do not believe many are purely mechanical. IMO it is not easy to provide a generic isolated interface for them.

I did a quick google of "lathe spindle encoder" there is a bewildering array of different styles...

I do not what kind of inputs commercial controllers provides for spindle encoders. I use a [SmoothStepper](https://warp9td.com/index.php/documentation/doc-ess) with Mach3 - and I had to make a BOB for that as there is no optocouplers on-board. If I want to add an encoder I have to provide a level-shifter and/or isolation if so required.

**Comment by terjeio** [comment]
A nice device to put on the encoder inputs could be the 74LVC1G17, a Schmitt trigger with 5.5V tolerant inputs. In my chase to track down my lathe problem I made a new encoder design using them.

The lathe problem turned out down to be the power supply, I was triggering the overcurrent protection randomly during cutting moves, but not during rapids - losing (or gaining?) steps. Reducing stepper current a bit solved that and subsequent tests of G33 spindle synced cuts lined up perfectly.

A bit more work on the lathe project and I will be ready to spend some time on the T4 driver and the sender. It would be nice to get spindle sync and Ethernet implemented for the T4.

**Comment by phil-barrett** [comment]
I love tiny logic and have an OR gate on the new board. Will have to take a look, see my comments are the end of this message. Just finished up the T4.1 board including the encoder inputs.  Here's what I have so far on that. Schematic first. Inputs on the pin header to the left. Connection Teensy pins on the right. Ignore the 10K value on the resistors, will probably be 330.
![enc input sch](https://user-images.githubusercontent.com/16570958/82156375-a5287b80-982f-11ea-85a4-ddbe9a4c207c.png)
The header on the board:
![enc input](https://user-images.githubusercontent.com/16570958/82156465-f9cbf680-982f-11ea-863b-062d1c06c5f7.png)
Voltage config via a solder jumper on the bottom side of the board:
![enc input v conf](https://user-images.githubusercontent.com/16570958/82156486-0f412080-9830-11ea-92bf-a17dfd3b6b5d.png)

Adding 3 schmitt triggers isn't hard. Would have to put it between the pin header and the opto inputs.  I'm not sure I see the benefit, though. An ST is good for slow moving analog signals. I think encoder outputs are fairly crisp. Is that not correct?  STs also help with noise though optoisolators do too. I've got a couple of different encoders here. Will take a look with an O'scope to see how they behave.

**Comment by terjeio** [comment]
What is the RL on the optocouplers? Only MCU input pullup? If so the frequency response will likely be really bad... See page 12 of the [datasheet](https://www.sparkfun.com/datasheets/Components/LTV-8x6.pdf). MCU input pullup resistor values are typically around 50K - not sure what the typical IMXRT1062 value is, didn't find it after a quick check.

The output of my spindle encoder is (or rather was) close to a sine wave, it is ok too feed that to a MCU input if it has a Schmitt trigger embedded. MSP432 has as standard, for IMXRT1062 it seems to be optional and configurable, ref 11.4.2.1: "Selectable Schmitt trigger or CMOS input mode". Something to keep in mind when adding driver support.

Since you decided to keep the opto-isolators it would be helpful to specify the max usable pulse rate. And also the minimum drive current required for the diodes unless you decide to put a Schmitt trigger in front of them - if you do so you do not load the encoder outputs too hard and can control the diode current yourself.

**Comment by phil-barrett** [comment]
OK, so maybe that's not going to fly. The 1062 has configurable pullups but they are pretty weak. I'm worried about a noise path into the digital section of the system. Maybe a low pass filter with an appropriately chosen cut-off frequency would work. Figure 50 KHz range?

Good point on the 1062's configurable inputs for ST - I will need to spend more time on that.

**Comment by terjeio** [comment]
> OK, so maybe that's not going to fly.

Well, just adding a resistor (in parallel with the pullup) will help a lot - then you may keep the optos.
Here is a PWM to voltage converter I made for the spindle VFD I use for the lathe:

![bilde](https://user-images.githubusercontent.com/20260062/82229340-996daf80-992a-11ea-91bb-21661a8136ec.png)

As you can see I drive both the photodiode and transistor quite hard, this in order to get a acceptable frequency response.

Still I am of the opinion that 74LVC1G17 on the inputs and no optocouplers would be a good compromise. Just add a pullup resistor on the input Schmitt trigger in case the encoder output is open collector/open drain.  IMO leave it to end user to provide additional circuitry if required. But do as you please, do not listen too much to me as I do not know what the appropriate choice will be...

I took a quick look at the spec of some spindle encoders I googled - I see some have differential outputs. Nice to have in a noisy environment.

**Comment by phil-barrett** [comment]
Since I've never used a spindle encoder on a motor, I don't have a good sense of the requirements. In the land of the blind...

Anyway, I'll probably use a low pass filter with a 30-50KHz or so cut off freq. I see the point of schmitt triggers here though. Considering setting up 4 ST inputs for general use.  May use 
 2 SN74LVC2G17s - a dual ST buffer with 5V tolerant inputs. The user can select a capacitor to meet their input frequency requirements or leave it off. 

I still need to do more research on encoder inputs in CNC work.

**Comment by terjeio** [comment]
> I still need to do more research on encoder inputs in CNC work.

Same for me. A good reason for not complicating this too much initially?
FYI for IMXRT1062 filtering can be done in hardware, se section _55.8.3 Input Filter Register (FILT)_.

The number users that will need the encoder input for a lathe is low to none? Perhaps making use of it as a MPG input as well could be attractive? Use the index input pin to toggle between the axes? I could add a plugin for that.

**Comment by phil-barrett** [comment]
Yes, an MPG is fairly high on my list. However, I'm thinking the Host USB interface on the T4.1 is more attractive as it can support all those USB based MGPs out there. I think that's more work but ultimately a better solution.  I've been meaning to talk to you about it. For the hardware, all I need to do is add a USB connector. Then it's a "small" matter of a grblHAL driver. With an ethernet interface, the PC may not be right next to the mill/lathe/... so an MPG plugged into grbl makes a lot of sense. On my list to investigate.

Also, I have looked at using an MPG via the keypad interface.  I've programmed a Teensy 3.2 (could even be a Nano...) to read the encoder and buttons of an MPG I have and act as an I2C slave to send encoded "keys" to grblHAL.  It's a bit rough at this point and was more of a learning exercise for me but it works.

Anyway, to get back to the input issue.  I've got 2 2G17s and low pass rc filters for 4 schmitt trigger inputs designed in. Will go with that for now as I want to get moving on the ethernet interface. If it's not right, c'est la vie.

By the way, which version of the 1062 manual are you looking at?

**Comment by terjeio** [comment]
> By the way, which version of the 1062 manual are you looking at?

Document Number: IMXRT1060RM Rev. 1, 12/2018

> ... act as an I2C slave to send encoded "keys" to grblHAL.

FYI: I am going to change the strobe input to active low with pullup so it can be wire-or'ed. This means that the pin driving this input should preferably be open drain - never drive it high.

**Comment by einencool** [comment]
Hello, today I finally got the Teensy 4.1.
And then I remembered a software called „EstlCam“  and there are some special features that I wanted to ask, if it is possible to add in a simple way.
It has a Potentiometer to set the Speed/Feedrate of the Program without pressing some Software Buttons. Especially in the beginning of a new GCode, sometimes it would be great to lower the speed of the machine, and if the first moves are ok, then the speed can be reraised to 100%.

Maybe you say: „Oh, that‘s very easy“ :-)
https://youtu.be/RE86ogmF0Vs

**Comment by phil-barrett** [comment]
Watched the video. Mein gesprochenes Deutsch ist nicht serh gut. Did not understand all he said but... The panel with the two buttons and control knobs is separate from estlcam. It probably plugs into the sender PC.  I think most senders can override feed rate in the UI. Spindle speed is possible from the VFD (or other speed controller), if present.  It would be possible build a panel like that as a plug-in to grblHAL. It would be run by a small microcontroller (even an arduino nano could do it) that sends messages via the keypad interface. Essentially, it's an MPG/Pendant type device.

**Comment by einencool** [comment]
Hi Phil,
Wow, I didn‘t know that you can understand and write in German, Respect 👍 

This Panel plugs directly to the Arduino Mega or the Arduino Nano. The Arduino Uno doesn‘t have the right pins with an analog signal input.

Here you can see the wiring
http://www.rc-network.de/forum/showthread.php/719360-Bedienpanel-ESTLCAM-Verkabelung

Drehzahl = Spindle speed
Vorschub = Feed Rate
And the two buttons with Start and Stop

**Comment by phil-barrett** [comment]
Ich versuche es, aber dein Englisch ist besser als mein Deutsch...

Yes, it took me a little time to understand that Estlcam is a complete controller too.

The stop and start buttons are directly available in grblHAL (and plain grbl, too). Drehzahl and Vorschub are not. So something would have to send those commands to grbl (or the G-Code sender) or the spindle controller (VFD or similar) - that's the MPG/Pendant approach I mentioned above.  Terje's GCode Sender does this pretty well (as do other senders). I prefer that approach - less hardware to fail.

So, it's possible but not terribly easy.

**Comment by terjeio** [comment]
The only way to change the spindle speed or feed rate for grbl on the fly while running a program is to issue real-time override commands. I _could_ add shortcuts for the buttons in my sender, but not sure this is a good idea as there are 13 of them... Any ideas for that?

I have a "potentiometer" (I am repurposing _Menu_ encoder for that when the DRO canvas is visible) for setting the spindle speed from [my MPG](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) prior to starting a job, but I have not yet implemented any overrides. _Cycle start_ and _Feed hold_ has physical inputs so easy to wire up. 

The [I2C keypad plugin](https://github.com/terjeio/grblHAL/tree/master/plugins/I2C%20keypad) can be extended to support real time changes of the overrides even when a job is running. _Flood_ and _Mist_ overrides are already assigned keycodes.

So there are a few options available for expansion, even a secondary input stream could be used if limited to real-time commands when grblHAL is not in "MPG mode."

"MPG mode" is a grblHAL concept - a secondary input stream is needed as is an input pin for switching MPG mode on/off. When on the sender UI is disabled for input and passively listen to the data stream in order to keep itself updated. More info at the link above.

A commercial MPG could use the MPG mode concept with an appropriate firmware driver.

Note that I have not yet added MPG mode to the Teensy driver.

**Comment by einencool** [comment]
Ok, i have to understand ;-)
I just thought, that it would be a great feature to start the job with a low feedrate, and when everything works fine, just bring it back to 100%.
The spindle speed is for me not necessary, and the Start, Hold and E-Stop I’ve already wired up.

In our company we have a 3d CMM and there we also use the feed rate adjustment every time...
I thought with this peace of hardware it could be very easy to apply, but it seems there is a lot more to do than i thought 💯

**Comment by terjeio** [comment]
> ... but it seems there is a lot more to do than i thought

Not neccesarily, but it needs to be done in some way or another. I have in fact updated the grblHAL core for accepting a quick succession of override commands in order to support potentiometer style adjustments by [buffering them](https://github.com/terjeio/grblHAL/blob/master/grbl/override.c). The original grbl code does not. I did this for my MPG but it implementing it there is on some long forgotten todo list.

**Comment by einencool** [comment]
Hi Phil,
Did you have the time to make some tests with the Teensy 4.1?
Do you think it would be possible to add the potentiometer for the feed rate adjustment?
In my opinion it would be a great feature for the machine and would be also a big advantage to choose the Teensy and the grblhal :-)

**Comment by terjeio** [comment]
@einencool : I think it is better to use the spindle encoder input, analog input from a potmeter is IMO not a good choice. Encoder input may also easier to handle since it will be used to calculate which (relative) feed rate override commands has to be sent to grbl. And no analog noise to worry about. 

Also, encoder input is easy to make relative to the current programmed feedrate - potmeter input is not. E.g. if feed rate override is canceled/changed by the sender what then to do with the potmeter? Move it to the appropriate position by using a motorized version?

When I find time I wil try using the _Menu_ encoder on my MPG for overriding feed rate _and_ spindle speed, encoders that has an additional switch (mine has) may be used to toggle between which override to change. The encoder index input can be used for that.

**Comment by phil-barrett** [comment]
Yes, I just got my 4.1 PCBs from China and will be building it up this AM.  I have 4 digital inputs fed through 5V tolerant Schmidt Triggers. Currently the plan is for a 1Kohm/10nF RC filter which gives a cutoff frequency of 15KHz. Alternatively, I could easily do 48KHz (330ohm/10nF) or 1.5KHz (1Kohm/100nF).  I could even leave off the caps if filtering is undesirable - it probably won't matter either way for manual input encoder.

And, yes, an encoder is the way to go on this.  Not only is it more precise (potentiometers, especially low cost ones, are notoriously imprecise) but it is simpler and doesn't have to deal with noisy ADCs in a very noisy environment.  Encoders with clicks are highly repeatable repeatable.

**Comment by phil-barrett** [comment]
A bit more updates on the T4.1 based design. It includes an ethernet jack which I hope to be testing soon. 

I have been looking into manufacturing a PCB with all the surface mount components and it looks good.  Found a manufacturer that can do small runs for a reasonable cost. I've had to change a number of components to work with what the manufacturer has available and will be doing a trial run of 10, probably next week. The design is ready to go once I have verified the PCB, A user would install the through hole components (headers and terminals mostly) and the T4.1
. I'm targeting less than $30 for the "unkit" board.

**Comment by phil-barrett** [comment]
ugh.  missing parts to assemble.  I modified the board to use smaller footprint capacitors and had ordered them a while ago. Unfortunately the package won't get delivered until Tuesday (not sure why it was so slow).  I can make some progress but full eval won't be done until then,

**Comment by einencool** [comment]
Thank you two for the fast responses :-)
As you know, my knowledge about electronics is not so good. When someone tell me, what to do, then it's not the Problem to solder some parts together :-), but like the digitl inputs, I have NOOO idea what I should do with it *g*

But it would be great, when it would be possible to adjust the rates with the Hardware which I can install directly to the machine, without going to my laptop which is standing 2 meters next to the machine.

Last week I wired up the Teensy 4.0 Board completly, with the vacuum cleaner and the mist coolant automated with the machine. That's great, and after the program is finished, I switch the button and can clean the machine with the vacuum cleaner. It works really good.

In the next time I'll try a little bit with the probing and maybe someday with the M6 commands.

Thank you both for all your great work

**Comment by phil-barrett** [comment]
Wunderbar! glad it's working for you.  Would love to hear feedback on any changes or ideas for improvement.

**Comment by einencool** [comment]
Well, for now with the teensy 4.0 it’s all ok. You mentioned that you make a improvement for the switch for the vacuum cleaner. Thats one point for easier wiring. The A-Axis maybe can be used in the future for the second Y-Motor, but that’s nothing to change...

Then for me it would be great with the feedrate adjustment (in a way you both can support without problems ) and the network connection. Fortunately I’ll have to swap my laptop because it doesn’t have a network connection *g*
I’m not so firm with the hardware and i don’t know, what is possible with the teensy, so if there would be a nice feature you can tell, let me know ;-)

**Comment by terjeio** [comment]
> The A-Axis maybe can be used in the future for the second Y-Motor

I have ported [grbl-sim](https://github.com/grbl/grbl-sim) to grblHAL (as a driver) with the intention to expand it into a test platform for features that is hard to test without access to a suitably configured machine. Auto squaring and probing are areas where this would be useful. I will publish the simulator with the next release of the grblHAL core, cannot do it for the current since I have made some optimizations to the planner that I need for the port. And also a fix for a bug that was obscured by another bug (both inherited from vanilla grbl) I fixed a long time ago that the port revealed...

One new feature for grblHAL that is now on my todo list is [probe connected detection and protection](https://github.com/terjeio/grblHAL/issues/54#issuecomment-634257797) - this could become a useful feature.

**Comment by phil-barrett** [comment]
That is interesting. I looked at grbl-sim. It makes a lot of sense to be able to simulate various CNC machines.

**Comment by phil-barrett** [comment]
Some good news, the ethernet jack on the PCB works with the T4.1. Some not so good news, the lwip port is rough and doesn't use RTOS.

**Comment by terjeio** [comment]
> Some not so good news, the lwip port is rough and doesn't use RTOS

A RTOS is not needed, polling lwip can be done from the systick ISR, from a TI example:

```
//*****************************************************************************
//
// The interrupt handler for the SysTick interrupt.
//
//*****************************************************************************
void
SysTick_Handler(void)
{
    //
    // Call the lwIP timer handler.
    //
    lwIPTimer(SYSTICKMS);

    //
    // Tell the application to change the state of the LED (in other words
    // blink).
    //
    g_bLED = true;
}
```
Note that the lwIPTimer fuction is specific to the TI lwip port (port = lowlevel driver), a similar function is available in the T4.1 port?

It should be possible to use the MSP432E401Y [implementation](https://github.com/terjeio/grblHAL/blob/master/drivers/MSP432E401Y/ethernet/enet.c) as a template for getting lwip up and running in grblHAL.

And if the [networking plugin](https://github.com/terjeio/grblHAL/tree/master/plugins/networking) is used for the protocol layer \(telnet and/or websockets\) then code can be copied from the  MSP432E401Y [driver](https://github.com/terjeio/grblHAL/blob/master/drivers/MSP432E401Y/base/driver.c) . Most of the code needed is wrapped in `ETHERNET_ENABLE`, `TELNET_ENABLE` and `WEBSOCKET_ENABLE` #if #endif blocks.

**Comment by phil-barrett** [comment]
Thanks. I'll probably be asking questions once I get farther into it.

**Comment by einencool** [comment]
I don’t want to open a new issue for this question...

It‘s not quite a Teensy 4.1 question, because I have this issue already with my „old“ Teensy 4.0 Board.
I have changed my stepper drivers/motors to Closed Loop JMC IHSS60-36 Steppers with integrated Encoder / driver etc...

With the old Motors I had limited the machine speed to 3m/min
The new Steppers were set to 6400 steps / rev with a leadscrew pitch of 5mm/rev —> 1280 steps/mm
Everything is working fine.

When I set the Limit to 5m/min the Steppers start to move and then it get‘s an error end makes a big noise (like you crash against a wall) This I can repeat, it‘s all the time the same.
It is on my X and Y axis (Z-Axis is only set to 2m/min) 
On the Y-Axis I‘ve installed 2 Motors and without them I can easily move the Gantry, the same for the X-Axis.

So I don‘t know what could be the Problem. Is it possible, that the Teensy didn‘t send the steps fast enough, or are they so fast, that the steppers didn‘t recognize them correctly? 

I set the Acceleration to 1000mm/s2 But also on 500 it doesn‘t work.

Tomorrow I‘ll test to set the Steppers to 2000steps/rev —>400 steps/mm
That should still be accurate enough for me, when it will work..

**Comment by phil-barrett** [comment]
1000 mm/s2 is pretty high acceleration. I've been running a lot lower.  50 mm/s2 seems like a sweet spot for my steppers though they are pretty cheap.  I run this on GRBL as well as grblHAL.

I'm looking at step rate timings on my scope right now. playing with high acceleration and step rates.  unfortunately, it's a new scope and I need to read the manual.  Will report back what I find.

**Comment by terjeio** [comment]
@einencool Had to tell what the problem could be, the stepper drivers should handle the step rate wich is about 106 KHz - max is 200KHz. Did you reduce the step pulse width? It defaults to 10 µs, this cannot be used for such a high step rate. Perhaps around 5 µs will be ok?

**Comment by phil-barrett** [comment]
A quick note, the waveforms coming out of the HCT541 look pretty clean. A small amount of ground bounce but better than a lot of others I've seen. Still fumbling with my scope...

**Comment by einencool** [comment]
@phil-barrett 
When I build the machine, I started with 100mm/s2 then 1000 and also 2000. It works all fine, but at 2000mm/s2 I was a little bit scared *g* Some people here drive there 600x400 machines with 5000mm/sec2 So normally that shouldn‘t be a problem. (I hope so...)

@terjeio 
That sounds plausible. Where do I have to change these values? Can I do it directly in the Software or do I have to build a new Firmware?
Tomorrow I‘ll try it with lower Microstepping, and then it should work, when the signal length should be ok...
How much steps could I use so that the 10my sek should be ok?

**Comment by terjeio** [comment]
Step width needs to be less than 10 - at 10 and 100KHz step frequency output could be glitches only. The steppers are rated for 200KHz, at a 50% duty cycle that is $10 = 2.5, 2.5 is not possible so either 2 or 3 is the minimum value that is acceptable. IMO you should not loose steps with $0 >= 3 (and <=7?).

**Comment by einencool** [comment]
The Datasheet for this steppers say that the step width and Pause should be more than 2,5my, so I think 5my should be ok, but I‘ll try it with 7 to be on the safe side *g*

I‘ll try tomorrow and report back

**Comment by phil-barrett** [comment]
With a 5 uS pulse width ($0=5), I was able to get a 150KHz pulse train.  Here's the screen shot of it.  It says 148.8kHz. I tried to capture fairly far into the move but there may be some residual acceleration going on. Still learning my O'scope. 

One other issue  that may not be important is the pulse width looks to be slightly off - 5.2 uS vs 5. See the second screenshot. I need to verify if the Teensy's clock is correct.
![ScreenImg](https://user-images.githubusercontent.com/16570958/86496035-693e6c80-bd30-11ea-9a70-b89b646b7ba4.png)

![150KHz](https://user-images.githubusercontent.com/16570958/86495805-b40bb480-bd2f-11ea-8c8f-c6f4f6308d50.png)

**Comment by einencool** [comment]
Can you say how long the pause between the two steps is?
The Datasheet says not less than 2.5my
![5562B22E-E287-4EDA-93B0-DDC8D19638E1](https://user-images.githubusercontent.com/37541729/86496764-117b2200-bd7f-11ea-8f78-a6051d11669b.jpeg)

**Comment by phil-barrett** [comment]
Hmmm, my response seems to have fallen into a black hole.  Anyway, what I tried to post a while ago:

The 2 pictures are from the same capture so the low period is 1.22 uS  (6.72 - 5.2) so it out of spec for your driver.  The minimum grblHAL pulse width is 3 uS so 5.5 uS period (or 181.818 kHz) is the fastest it can go for you.

**Comment by phil-barrett** [comment]
More info on this.  I played with pushing the frequency higher.  Things seem to fall apart above 164kHz.  Here's a screen shot at 163.9kHz. A fair amount of ringing but it will drive an opto OK. At 180, I am seeing very strange behavior including oscillations and short (500 nS) pulses.  My Oscope (cheap probes and my lack of familiarity with it) may be partly to blame here but it's got 100 MHZ bandwidth.

By the way, with the step pulse set at 3 uS, I am seeing 3.2 uS on the scope.  So, somewhere an extra 200 nS is getting added.  Again, I don't think it's a huge issue.
![163 9khz](https://user-images.githubusercontent.com/16570958/86505826-3cb14180-bd7e-11ea-9e79-0e8c63a40c6e.png)

**Comment by terjeio** [comment]
> Things seem to fall apart above 164kHz.

The signal is breaking up? If chasing high step rates perhaps USB comms is not the best option, better to use a USB <> UART converter?

>  So, somewhere an extra 200 nS is getting added.

Some I believe is due to interrupt latency, can be tuned by subtracting a bit from `TMR4_COMP20` value assigned from `settings->steppers.pulse_microseconds` (around line 700 in driver.c);

The pulse width setting should be changed to a float and the lower limit changed a bit. This to allow more fine grained control at high step rates - remember the original grbl is limited to around 30Khz...

One change I have wanted to do for a while is to only introduce the step delay on a direction change, IMO no need to do that continuously and thus wasting processor cycles. A direction change cannot be done abrubtly so a few microseconds delay on changes should not matter.

@einencool Be aware that your driver specifies a minimum delay of 5 µs for the first step pulse after a direction change, grblHAL defaults to 0 - set by the $29 parameter. Perhaps 5 µs should be the default value for this parameter?

**Comment by einencool** [comment]
Ah, thank you for this hint, I think it would be a good point to set it as default

**Comment by phil-barrett** [comment]
Thanks. that makes sense.  I'll have a look but think part of the problem is the use of CMOS logic and relatively long traces from pin output to connector. Neither are going to change any time soon...  I don't know if there is much need to go above 160kHz 

The stepper drivers I looked at (including Geckos, both new and older) list 200 kHz as their max frequency, so it is a potential issue. Maybe for a follow on product...

I think your ideas make sense and should be done (step delay after dir change and float for pulse width). 

One thing I will do at some point is to take a look at what's coming out of an unloaded Teensy to make sure the problem isn't in grblHAL.

In the words of a bad American television car commercial - this isn't your father's GRBL anymore!

**Comment by einencool** [comment]
Today I haven‘t the time to try the different settings. 
So I‘ll try them in Monday. 
But you are all right, I will change the Steps/mm so that there will be no issue. 
The Motors got the Settings 6400 Steps/rev from the beginning, but then I‘ll change them down, that not a big deal, But for two Motors I‘ll have to disassemble half of the machine *g* So for the next time I‘ll stay with 3m/min because I also have some trouble with the Motors.
When I plug them to the source, they start to make a high freuquency noise, like a capacitor is loading....
These are all 3 Motors I got from the manufacturer from China and all make the same noise.
Today I got a Motor from a german supplier and this is making no noise... Very strange...

But let me ask another question for the new Teensy Board.
For my old machine I build a Z-Probe which will be connected to the Probe Pin and with the Ground on the Spindle it will measure the tool length. (This method uses a „normally Open“ Contact, but doesn‘t have a real „switch“ inside)
Now I got a 3D-Finder which has a „normally Close“ Contact.
I also have another Tool Probe which also has a Normally Closed Contact, but the Measurement surface is really bad (0,25mm on a circle of 20mm) And it seems, that I can‘t disassemble this thing.
So would it be possible to set one Modul on the „Probe Pin“ and the other Modul maybe on a not used End-Stop input (B-Axis) ??
I‘ll never use a B-Axis, but the A-Axis Endstop may be used in the future for autosquaring the Y-Axis...
Or do you know a way, I could use the different Probes? :-)

**Comment by phil-barrett** [comment]
Basically, you want to have multiple probes, I believe.  In terms of the hardware, you could use an end stop or one of the digital inputs. The issue is the software using it. I would look to see if it is possible to create a second probe input or switch the probe pin on the fly with a plug-in (maybe a bad idea).

On the motor whine - perhaps the Germans have higher standards than the Chinese?

**Comment by einencool** [comment]
I’m looking for a new touch probe. Sadly i have no ATC Spindle, that would make this things so much easier *g*

The Motors are all from „JMC“ in China. And they both have the same „Version“ 1.45
All the DIP Switches are set to the same value. Tomorrow I’ll make a video and send it to the manufacturer...
In the same time I’ll read the parameters from both devices, maybe there is a difference...
So for now the motors directly from china are way more scrappy than the same one from germany, obviously they are the same...

**Comment by einencool** [comment]
With 3200 Stepd/rev I was able to set the Speed to around 7m/min (I haven‘t tested more, because that‘s absolutely enough, I think I‘ll lower it to 5m/min :-) )

Now I have two questions to the „Pros“ :-)
I ordered another Touch-Probe with a normally Closed switch, I hope it will arrive in the next days.
But I have a little Problem with my thinking. I have two Systems for this Machine: 3D-Finder for Edge Finding and maybe Z-Probing / and the Z-Touchprobe.

The 3D Finder has a Power Connection and the Switch is normally Close (So far, so good)
The Z-Touchprobe also uses only a NC-Switch

In my Thinking it would be great, to plug in the 3D-Finder, when I need it, otherwise I‘ll disconnect it and put it in a case. 
The Z-Touchprobe can be connected the whole time, or I‘ll make also a Plug on it.

What do you think would be the best Option? Connect only one System at a time, or make it somehow possible, to connect the 3D-Finder when I need it, and let the Touchprobe connected to the machine so that both Systems should work „in a Row“?

The other question is for the encoder on the Teensy 4.1
How must it be connected and configured so that I‘m able to set the feed rate?
With the Push-Button on the Encoder, would it be possible to set the feedrate back to 100%?

Thanks for your input :-)

**Comment by phil-barrett** [comment]
On the second question: I've added a [zip file with configurations to the github repository](https://github.com/phil-barrett/grbl-teensy-4/blob/master/configurations.zip).  Look at the "5 axis quad enc" directory - driver.h includes pin assignments.  On resetting feedrate to the default, you would need to change encoder.c to make that happen. Probably fairly simple but I don't have time to look at it. By the way, I have done a fair number of updates and additions to the docs in that repository - they are all linked to [the readme file.](https://github.com/phil-barrett/grbl-teensy-4/blob/master/README.md)

On the first question,  if I understand correctly, you want to be able easily switch between the two.  Perhaps a simple solution is to use something like an audio jack that switches sources when you plug it in. Essentially, plugging the second one in disconnects the first one. I don't know if you would need to make changes to grblHAL configuration.

**Comment by einencool** [comment]
Then I‘ll have a look at your Configs and your updated Docs tomorrow. Thank you

For the Touch Probes, I think I‘ll use only one at a time. Then I‘ll use the Mini XLR Connectors which was build onto the 3D-Finder. I ordered some Jacks and will connect them.

**Comment by terjeio** [comment]
For handling multiple probes a bit of electronics could be used; a transistor and a few resistors to invert one signal and a small relay to select between them. The relay can be controlled from g-code by a M64 command. Or use a switch instead of a relay.

The encoder input needs pins to be assigned, currently there is no functionality for resetting to 100% other than turning the encoder. The encoder code is rather crude, my plan is to expand it. Setting for sensitivity (number of pulses per change) is required, resetting to 100% could be another change - perhaps by click or double click. It should be possible to toggle between feed rate and spindle speed so I will keep that.

I have now completed maintenance work inside my house, but still have some left outside (painting).  So I have some time available for coding again - at least when it is raining as it is today...

**Comment by einencool** [comment]
Today I tested the new 3D-Finder with the Teensy 4.0 Board. It seems to work good, and I‘ll just disconnect it after Probing. Today I didn‘t make any parts, so, when I disconnect the Probe, the Red light for the Probe Signal is lightning. But it‘s almost possible to run the machine without it I think, or did I miss anything?
The Tool Length Sensor didn‘t arrive today, So I hope I‘ll start to change the PCB tomorrow. 
Today I got a little rework of the parts I make in the past days...

**Comment by terjeio** [comment]
> Today I didn‘t make any parts, so, when I disconnect the Probe, the Red light for the Probe Signal is lightning. But it‘s almost possible to run the machine without it I think, or did I miss anything?

The probe status is ignored during normal operation so can be safely ignored.

**Comment by einencool** [comment]
> On the second question: I've added a [zip file with configurations to the github repository](https://github.com/phil-barrett/grbl-teensy-4/blob/master/configurations.zip). Look at the "5 axis quad enc" directory - driver.h includes pin assignments. On resetting feedrate to the default, you would need to change encoder.c to make that happen. Probably fairly simple but I don't have time to look at it. By the way, I have done a fair number of updates and additions to the docs in that repository - they are all linked to [the readme file.](https://github.com/phil-barrett/grbl-teensy-4/blob/master/README.md)
> 
> On the first question, if I understand correctly, you want to be able easily switch between the two. Perhaps a simple solution is to use something like an audio jack that switches sources when you plug it in. Essentially, plugging the second one in disconnects the first one. I don't know if you would need to make changes to grblHAL configuration.

Hi @phil-barrett 
I downloaded your Configs, and it seems that you set the Encoder Plugin in the 3-Axis WITH Encoder to "0"
#define QEI_ENABLE         0

Maybe you have fixed it in the last 2 days, but that's what I see in the Moment :-)

**Comment by phil-barrett** [comment]
Sorry about that. I've been tweaking that and must have made a mistake.  will correct this morning and let you know.

your pin assignments look ok.  As long as you don't reverse voltage, there should be no harm.

**Comment by phil-barrett** [comment]
Sorry, it will say QEnc in $i. Maybe I should change that to QEI to match the code.
[VER:1.1f(IMXRT1062).20200603:]
[OPT:VNMSL,35,1024,5,0]
[NEWOPT:ES,TC]
[DRIVER VERSION:200606]
[DRIVER OPTIONS:USB.2 IOPORTS QEnc ]

**Comment by phil-barrett** [comment]
OK, I fixed the configurations.zip file, verified it is correct and uploaded.  Note that I have changed the name of the repository to grblHAL-teensy-4.x

Sorry for the errors before.  I will be more careful in the future.

**Comment by einencool** [comment]
:-)
Don't be so hard, I'm happy that I could help and find a issue, that's normally the other way *g*

It was just a hint, that there is a little mistake ;-)

**Comment by phil-barrett** [comment]
@einencool, I have been thinking about your desire to have multiple probes attached.  One way to do it is to use the Aux Relay output on the board to select the one you want to use. Since it has 5V TTL output, it's possible it could be used to enable a specific probe (assuming they are 5V tolerant/compatible. In a macro, you could use M64 and M65 to enable/disable.  

So if you have one probe that is enabled by Aux0, you would have an M64 P0 in a macro and that would set the AUX0 output to high (5V).  M65 P0 would set it to 0. The 5V output is limited to 8 mA.  If the device runs off of 12V you could use the relay coil outputs to power it (up to 100 mA).

**Comment by einencool** [comment]
Hmm, sounds interessting, Today the new Z-Probe arrives. It's only a NC Switch with these "6 balls" inside. The 3D-Finder needs also the 12V Power. 
But Iwill make a little Box in Front of the machine to Plug in the Probe I want to use.
I wanted to change the PCB today, but I ran into a little Hardware Problem with the new Z-Probe, because my Z-Axis travel was often at its Limits, so I decided to rise my complete Spindle Holder for around 45mm. Wtih the old Setup I was able to drive 20mm underneath my Wasteboard, so I just moved the Holder.
Now I was tramming the Spindle. It's now within 0,04mm on a 380mm Diameter, Thats close enough for me, because the Holder hast a little Bow in it and it drives me a little crazy, when I didn't set all the screws in the same way, then it moves like it wants....

But now I'll change the PCB and tomorrow I hope I'll attach the new Box in Front of the machine also with Switches for the Motor PSUs, the VFD and the ShopVac... And seriously the Encoder :-)

**Comment by phil-barrett** [comment]
yeah I hear you about tramming.  my current machine spindle trams with shims.  now that's frustration.

**Comment by einencool** [comment]
Oh, I have also to put Shims under my holder...
It‘s one of the cheap „Zink-Druckguss“ Parts, that was sold with the Spindle. It‘s fairly „OK“, but nothing more :-)
Last days a made a Plate which is installed to the Spindle Holder, where I can place the 3D-Finder. I made it with a Slot and a „V-Groove“ so that it will set everytime the same place and today I made a little test (before I changed the Spindle Position) and I set it into the V-Groove in different heights. The repeatability was around 0,05mm and when the 3D-Finder was in always in the same position it was around 0,004mm (I made around 5 tests in every Setup)

The 3D-Finder is fixed in a 150mm long ER20 Collet holder and this Holder is connected wih 3 strong magnets to the V-Groove. So I can attach the 3D-Finder within Seconds to my machine and I don‘t have to change my tool every time... I think That‘s pretty cool **g**

**Comment by einencool** [comment]
So, Today I Prewired the Machine and the Encoder runs fine. First it ran into the wrong direction, but after Changing the two wires, it works as expected, THANK you both for your supreme support :-) 🥇 

But I‘ve got one question. I‘ve a VFD Controlled 2.2kw Spindle and the RPM is controlled with the 5V PWM Signal.
I Powered the Teensy and the Board with the USB Power and a 12V PSU.
The Spindle should work in a range of 6.000 - 24.000 RPM So far, so good.
But I have the Problem, that the RPM is every time to low. 

If I set 6.000 RPM it‘s around 5.100 RPM 
With 12.000 it‘s around 10.500
And with 24.000 it‘s at 21.700 RPM

I haven‘t adjusted the 12V Poti to 10V, because in my opinion I don‘t need it.
Could be the Problem, that my Laptop doesn‘t reach 5V USB Power? So the Controller doesn‘t send the right Voltage to the VFD Controller?
When I attach a 5V Step Down to the Board, I only need to cut the 5V line under the PCB, right?
Tomorrow I‘m not working at the machine, so I can test it only on Saturday.

With the old Teensy 4.0 Board I worked with the 5V Step Down Board and the Spindle reaches the desired Speeds nearly on point...

Maybe you have a little hint for me, but I think it‘s a Problem with the 5V Power supply from my Laptop.

**Comment by phil-barrett** [comment]
I am a little unclear on a couple of points.  What voltage do you get when measuring the 5V screw terminal (and not other power than USB attached)?  Also, which output are you using with your VFD?  A or B in the picture below?
![PWM out](https://user-images.githubusercontent.com/16570958/87076161-308e0f80-c1d6-11ea-9b26-2536a4edcd25.png)

**Comment by einencool** [comment]
I was in rush today, so I didn‘t measure the Voltage from the USB *sadly*
I use Section „A“ for the VFD

**Comment by phil-barrett** [comment]
If you see 5V on the 5V terminal then it's not USB power.  I doubt that's the problem because your PC would be out of spec and other bad things would happen.

Is it possible you have the max spindle speed parameter in grbl ($30) different from the speed you requested? I was just looking at the spindle PWM on the oscilloscope yesterday and grblHAL is generating full PWM (solid 5V out on the PWM pin) for full speed (requested speed the same as $30). 

And yes, if you are using PWM to directly control the spindle, you don't need to adjust  R25.

By the way, if you have requested the same RPM as $30, you should be able to measure 5V on the PWM terminal in the "A" group.

**Comment by einencool** [comment]
I saved the Config from the Teensy 4.0 Board and they match perfectly.
I‘m writing from an iPad, so I don‘t have the values in front of me, but it was 
Spindle min 6.000
Spindle max 24.000
Voltage min 25%
Voltage max 100% 
And the Freuquenc was 4.000Hz / with 5.000Hz it doesn‘t match. And it‘s easier to change it in GRBL, than in the VFD Menu :-)
The Laptop was for quick testing without PSU, so I don‘t know if it lowers the Voltage to maybe 4.6V.
But I‘ll test it on Saturday, what could be the root cause..

But thank you so far for your input. If I have been a little bit smarter, then I would have checked the Voltage on the PWM Pin...

**Comment by phil-barrett** [comment]
This is strange.  The circuits on the two boards are exactly the same.  I'll pull out the old T4.0 based board and check it with the new Oscope to see if anything obvious stands out but there should be no differences.

When you say Freuquenc, I assume you mean $33 in grbl settings.  I will play with changing that and see what happens.

**Comment by einencool** [comment]
I think it must be $33.

Like I mentioned, I think it‘s a problem with my Laptop and not with your PCB

**Comment by phil-barrett** [comment]
To be honest, I think your laptop is not the problem.  

One simple thing to try.  Disconnect the PWM signal and connect 5V directly to the VFD PWM input. You can use the 5V terminal on the breakout board. Your VFD/Spindle should run at full speed - that is the same as 100% PWM. If it does, that rules out your VFD.  

Also, reconnect the PWM signal and try setting the speed to 90% to see what happens. If the VFD runs at 21.600, then there is something wrong with 100% PWM.

I looked at the PWM signal out on the scope a few minutes ago and it looked ok at 4 kHz (as well as 5 kHz). I have noticed a little 60 Hz (powerline, your's would 50 Hz) noise but it is fairly small and seems unlikely to be an issue.

**Comment by einencool** [comment]
Ok, I‘ll try it.
Thank you Phil for your continuous support, I‘m so thankful that you help me :-)

**Comment by phil-barrett** [comment]
No problem, I learn a lot in helping solve these kind of problems.

By the way, the more I think about, I am guessing your VFD does filtering on the PWM input similar to my circuit. So, a low board voltage (4.25 V in this case) would cause what you are seeing. Let me know what you find.

**Comment by einencool** [comment]
That‘s why I think that will be a Laptop Problem :-)
You can be sure, that I‘ll report back, what I found :-)

Thank you and have a nice day/evening

**Comment by einencool** [comment]
Sooo, now I'm in front of my machine and connected only the Laptop, and what should I say, it doesn't matter if I connect it with or without PSU, it spends only 4.48V out of the USB Plug...

So it's a crappy Thinkpad *g*

**Comment by phil-barrett** [comment]
OK. I am sorry and surprised to hear that. The USB voltage spec is +/- 5% so the minimum is 4.75V. I guess you need a stepdown for this too. Or, you could try running your VFD with the 0-10V output as that does not rely on having 5V (uses 12V).

You got me wondering about how widespread this is. I have a Sony Vaio and will check to see what it puts out. I may put a note in the manual about low USB voltage.

**Comment by einencool** [comment]
I installed the StepDown, that was used for the Teensy 4.0 Board..
Before that, I cut the line for the 5V, checked everything and then I powered up everything.

Then the Spindle speed worked as expected.
So I think a Hint in your Docs would be very good.
One opinion, for less Problems for future users, maybe to add a StepDown directly on the Board and only apply the 12V Source. I don‘t know if it‘s possible, but I think that would be a good Option, but it‘s only a thought...

Today my last missing Motor arrives and I‘ll wire it up on Monday...

**Comment by terjeio** [comment]
Is there a 5V output available on a terminal on the VFD that can be used? It is on mine and I did use that to ensure (I hope) that the PWM voltage matches the VFD. I did the same earlier for a brushless motor driver that had a 10V supply terminal and 0-10V speed control.

**Comment by terjeio** [comment]
@phil-barrett Have you started coding for ethernet? If not then I am going to add the needed code in driver.c as a start, then check if is possible to use lwIp port with the existing telnet/websocket protocol plugin.

I have added RS-485 ModBus code for the Huanyang VFDs to the Teensy driver - when the messaging is verified this will become another option for spindle control.

**Comment by phil-barrett** [comment]
I got distracted so no.  In typical PJRC fashion, finding the actual working lwip T4.1 code has been like chasing my tail.  The most current is here: https://github.com/ddrown/teensy41_ethernet and not in PaulStoffregen's repository.

I have a Huanyang VFD on order so will take a look when I get it.  What transceiver are you using?

**Comment by terjeio** [comment]
Ok, hopefully the cards will arrive soon - then I'll give it a try. Thanks for the link.

Currently I am using a VFD emulator I wrote for a TM4C123 board so no tranceiver. I have not tested the code against a real VFD yet. Code supports both auto direction and direction selection via a GPIO pin.

**Comment by phil-barrett** [comment]
Sigh, I looked at the link I gave you (got it this morning so hadn't even looked).  It doesn't appear very done to me. The ftp lib appears finished but the lwip lib is on his todo list.

The problem is that the lwip code is in a bit of a mess.  I get errors building the apps like lwip_websrv.  Oddly enough, earlier I was able to build them and they worked on my board. I updated to a new version of Teensyduino and nothing compiles - I get parameter mismatches and such. Sigh.  I was hoping not to have to dig in and figure out their source tree.

**Comment by phil-barrett** [comment]
Well, I was wrong, it is working. I was able to compile with ddrown's repository.

**Comment by einencool** [comment]
Hello, 
Today I worked a little bit with my machine and then I wanted to test the new installed Encoder with the Teensy 4.1 Board. And here comes my first question. Is it possible to change the Feedrate for Rapid Moves AND Cutting Moves together? 
Because In my opinion I wanted to use it, to have the possibility to slow down the Movements from the machine at the beginning from a new Piece/Job. But the Problem is, that the Machine runs with 5 m/min to the point and if there would be something wrong, I can‘t react so fast. 
I have seen the „encoder Config“, but is there a file, where I can see, what variables I can use?

Right now the machine makes a really good job. And I can only say thank you so much. 🥇 

The only thing for me is now getting my 3D-Finder and the Toolchange Macro to work. 
The 3D-Finder works, but I need to figure out, how I can see the measured points, so I can redefine the Offset from the ruby tip and the Touch Point.

For the Toolchange, I‘m a little bit confused how I have to work with the Offsets.
In my Mind I would like the following steps:
- Measure XYZ with the 3D-Finder
- Touchoff on the Tool Length Sensor with 3D-Finder
- Touchoff Endmill on TLS (today with G10 L20 directly on the Workpiece)
- Calculate and set Offset
- Mill first part
- Change tool and Touch on TLS and set next offset
This could also be done with different Files so that I can use normal macros...
Did anybody have some good ideas?

**Comment by terjeio** [comment]
> Is it possible to change the Feedrate for Rapid Moves AND Cutting Moves together?

I am not sure it is a good idea, first the starting value is likely to be different and there are only a few steps available for rapids. IMO a better solution is to toggle between rapid and cutting feedrate and adjust them separately.

> I have seen the „encoder Config“, but is there a file, where I can see, what variables I can use?

If you mean the available adjustment range then this is defined in [grbl/config.c](https://github.com/terjeio/grblHAL/blob/master/grbl/config.h):

```
// Configure rapid, feed, and spindle override settings. These values define the max and min
// allowable override values and the coarse and fine increments per command received. Please
// note the allowable values in the descriptions following each define.
#define DEFAULT_FEED_OVERRIDE           100 // 100%. Don't change this value.
#define MAX_FEED_RATE_OVERRIDE          200 // Percent of programmed feed rate (100-255). Usually 120% or 200%
#define MIN_FEED_RATE_OVERRIDE           10 // Percent of programmed feed rate (1-100). Usually 50% or 1%
#define FEED_OVERRIDE_COARSE_INCREMENT   10 // (1-99). Usually 10%.
#define FEED_OVERRIDE_FINE_INCREMENT      1 // (1-99). Usually 1%.

#define DEFAULT_RAPID_OVERRIDE  100 // 100%. Don't change this value.
#define RAPID_OVERRIDE_MEDIUM    50 // Percent of rapid (1-99). Usually 50%.
#define RAPID_OVERRIDE_LOW       25 // Percent of rapid (1-99). Usually 25%.
// #define RAPID_OVERRIDE_EXTRA_LOW 5 // *NOT SUPPORTED* Percent of rapid (1-99). Usually 5%.
```
Currently there is no config setting for encoder sensitivity, that is how big the adjustment will be for one turn on the encoder. This will likely be added in the next release - I am thinking of adding settings for a number of encoders, perhaps up to 5?

Regarding tool length offset I hope somebody with a good understanding of the ins and outs of this can chime in.

**Comment by einencool** [comment]
Hmm, that's not so good, but you're right.
Especially at the beginning of a job I'm often not sure if everything works fine. 

So I'll try in the next days if I can modify the Post processor that all G0 commands will be replaced with G1 F5000 
Then I could change it like I would like to have it.

Does GRBLHAL and your Sender work with commands like
G1 F5000 X50 Y50
Or must be the feed rate behind the command?
Especially in the Fusion PP programming I'm not so firm, because I'm not a coder :-)

**Comment by terjeio** [comment]
> Hmm, that's not so good, but you're right.

See [this comment](https://github.com/terjeio/grblHAL/issues/73#issuecomment-659222664), if your encoder has a switch that is activated by pressing the axis (or you can add a separate push button) then an encoder defined as universal can function in three different modes.

> Does GRBLHAL and your Sender work with commands like
G1 F5000 X50 Y50
Or must be the feed rate behind the command?

Yes they will, and the order of the parameters on a line is irrelevant - the gcode standard (NIST) defines a strict order of execution for the different values. grbl conforms to that.

**Comment by einencool** [comment]
Hi Terje,
Then I would use the encoder in Mode 1, yesterday I changed the PP so that all G0 Moves are G1 Commands and tomorrow I’ll try to use it. When it works, it would be great when the button is pressed, that the program runs with 100%.

I also could change it in that way, that the Feedrate is at the end of the command.

**Comment by terjeio** [comment]
> The pulse width setting should be changed to a float and the lower limit changed a bit.

I have now done this for the T4 and subtracted 200 ns from the step length to account for latency. 

![bilde](https://user-images.githubusercontent.com/20260062/87847148-f60f2b80-c8d5-11ea-97a5-95d3850b8209.png)

The step signal breaks apart for me when the step rate crosses a bit over 300 kHz in a test with a single move commanded (no program running). This with a 2.5 µs pulse width. With a pulse width of 2 µs the breakup happens at about 400 kHz. This agrees with [a test I did initially](https://github.com/gnea/grbl/issues/67#issuecomment-586289165), IIRC this with a running gcode program.

> One change I have wanted to do for a while is to only introduce the step delay on a direction change ...

With AMASS enabled this is not neccesary? Initial testing shows me that the first step call after a direction change does not output any steps. So there is an implicit delay that IMO normally will be larger than the stepper driver requirement.

**Comment by terjeio** [comment]
@einencool This was a tricky one to track down, but I hope I have managed to do so now. I reduced the polling frequency for new data from the USB stack, now 1 KHz - before undefined (same as the main grbl protocol loop - perhaps in the 100s of KHz and variable). After I did this I no longer have had any random crashes.

I have not checked but I suspect the polling USB stack may disable interrups briefely, and if it does so this _could_ be the reason for the data loss. I will run a few more test before committing the fix to github - this unless the randomness resurfaces.

**Comment by terjeio** [comment]
A fix is up. Note that I have changed Arduino based drivers to libraries, build procedure is changed! Hopefylly this does not break anything...

**Comment by phil-barrett** [comment]
Thank you.  Just to be clear, I believe this is the test branch only.  Master branch does not have the problem?

**Comment by einencool** [comment]
I Think so, because the problem occured first time with the test branch from 05.08.2020.
I never had Problems with the "master branch" and the first step from the encoder...

**Comment by einencool** [comment]
in the Encoder and eeprom Plugin there are little mistakes with the following lines:

#ifdef ARDUINO
#include "../../driver.h"

there is a ../ to much
#ifdef ARDUINO
#include "../driver.h"

maybe this is also in other plugins, but I only use these two at the moment.

Then I got two warnings, but it compiles: (but I don't have the Teensy here)
C:\Users\Einencool\Documents\Arduino\libraries\grblHAL_Teensy4\src\i2c.c:97:13: warning: 'force_clock' defined but not used [-Wunused-function]
 static bool force_clock (i2c_hardware_t *hardware)
             ^
C:\Users\Einencool\Documents\Arduino\libraries\grblHAL_Teensy4\src\encoder\encoder.c:103:13: warning: 'mpg_move_absolute' defined but not used [-Wunused-function]
 static bool mpg_move_absolute (uint_fast16_t state, axes_signals_t axes)
             ^
Opening Teensy Loader...

**Comment by terjeio** [comment]
> Master branch does not have the problem?

Technically it has the same problem - but for some reason it does not show up. Could be that the failure rate is just significantly lower? Perhaps it could/should be checked? When I lowered the step pulse interrupt priority to 1 the failure rate went to 100% for the few tests a ran with the test branch code so this could be an easy check? But since I spent the best part of yesterday struggling with this - and the master branch is soon to be updated this will be a waste of time?

**Comment by terjeio** [comment]
> in the Encoder and eeprom Plugin there are little mistakes with the following lines...

Ouch - I must have forgotten to sync the plugin code. Will do so asap. Sorry for that.

**Comment by einencool** [comment]
No Problem, these little Problems I'm able to fight with *g*

**Comment by phil-barrett** [comment]
OK, I will wait for the mast branch update to test. Getting ready to put the ethernet board in my machine so will be testing that soon.

**Comment by einencool** [comment]
That sounds good :-)

So, I updated the machine via Teamviewer.
I could only make a little test, without running the cnc, because my father can't do anything with this machine :-)

Normally the settings were reset after I flashed the device, but this time, the settings were the same as before.
Do I have to activate this in the config to clear the eeprom, or is it activated by default?

**Comment by terjeio** [comment]
> Normally the settings were reset after I flashed the device, but this time, the settings were the same as before.
Do I have to activate this in the config to clear the eeprom, or is it activated by default?

No need to do anything if they are intact. IIRC you have an EEPROM enabled so settings will survive a reflash unless the new version has changes in the settings data structures.

**Comment by einencool** [comment]
I was testing to usw the Teensy 4.1 Files for the Teensy 4.0 Board, without EEProm and Encoder and so on.
I was getting some errors, I think because of the different "Enable Pins". 
Sadly the CNC-Boosterpack doesn't fit to the Teensy4.0
So I copied your Teensy 4.1 Board File and changed some things, so that it would fit to the older Teensy 4.0 Board from Phil.
I also changed the Pins for Auto-Squaring to the A-Axis, but I can't do tests here, because the Board is only on my desk. But it compiles without errors, While I always get warnings with the Teensy 4.1 with Encoder and EEProm enabled, the rest is the same.

If it would help someone, attached is the "Teensy40_map.h" 
[Teensy40_map.h.txt](https://github.com/terjeio/grblHAL/files/5124903/Teensy40_map.h.txt)

and here is the addition for the "my_machine.h"
//#define BOARD_T41U5XBB
#define BOARD_Teensy_40
//#define BOARD_CNC_BOOSTERPACK

I don't know, if this would be the right way, and if everything is working as expected, but the board seems to work with your software :-)

**Comment by phil-barrett** [comment]
There are only a couple of 4 Axis boards out there (I didn't do an unkit for that one) but thanks for making that available.

**Comment by terjeio** [comment]
> I don't know, if this would be the right way, and if everything is working as expected, but the board seems to work with your software

Thanks, I'll rename the file to T40X101_map.h which I belive is more in line with the title from the schematic. Phil, is this ok/correct?

If you had commented out both boards generic_map.h would have been used which was the same mapping as the one that was used before. But better to have an explicit mapping file. I will add a link to Phils's github project in my_machine.h as well.

**Comment by phil-barrett** [comment]
Sorry, we crossed in the posting. That name is fine based on the schematic though following the naming convention (that came later) it would be T40U4XBB_map.h.  

This does make me wonder how I should handle later versions. In general I will keep the pin assignments the same though add new features.  Coming next is an update to the serial header (added +3.3V), Host USB (combined Ethernet and USB plug) and a couple of minor layout changes.

**Comment by einencool** [comment]
I‘m glad, that I could help a little bit with this :-)
I didn‘t know, that commenting out all two boards would take the „old“ settings, but ok, so I think it should also work, and the squaring is also available :-)

**Comment by terjeio** [comment]
> That name is fine based on the schematic though following the naming convention (that came later) it would be T40U4XBB_map.h.

If you are not going to change the schematic I will use T40X101.

> This does make me wonder how I should handle later versions. In general I will keep the pin assignments the same though add new features.

New features should be possible to enable from my_machine.h or from the *_map.h file if not optional. A new version = a new map file depending on the changes. I'll add a availability matrix for options in my_machine.h:

```
              Plugin: | ETHERNET¹ | SDCARD¹ | KEYPAD | EEPROM |
----------------------|-----------|---------|--------|--------|
BOARD_T40X101         | no        | no      | yes    | yes³   |
BOARD_T41U5XBB        | yes       | yes     | yes    | yes³   |
BOARD_CNC_BOOSTERPACK | yes²      | yes     | yes    | yes    |

¹ Teensy 4.1 only.
² External magjack.
³ EEPROM is optional and must be added to the board.
```
> I didn‘t know, that commenting out all two boards would take the „old“ settings

Not easy to know that this would match your board - so good that you made a new board file. Importing the project as a library was trouble free for you?

**Comment by phil-barrett** [comment]
I'm not changing that schematic - no more Teensy 4.0 development.

**Comment by einencool** [comment]
> >
> Not easy to know that this would match your board - so good that you made a new board file. Importing the project as a library was trouble free for you?

I found it a little bit confusing, because in the old versions the .ino file could be started directly. And all changes could be made in this directory. Now I need to move to the directory in the documents folder.
And I think I have to delete the old files before I can use it with the newer versions.
For me it's a step back, but I have to arrange with it :-)

**Comment by phil-barrett** [comment]
I believe this has been resolved to everyone's satisfaction. (understatement of the year?)
